### PR TITLE
fix(toby): retry tool calls without reasoning effort

### PR DESCRIPTION
**Pull request by OliverBryant**
## Summary

- retry OpenAI-compatible Chat Completions tool requests with `reasoning_effort="none"` when the provider explicitly requires it
- preserve function tools during the compatibility retry
- keep the retry bounded and leave unrelated bad requests unchanged

## Production reproduction

AWS Bedrock GPT-5.6 Sol rejects function tools on `/v1/chat/completions` while its default reasoning effort is active. A direct request through XAgent's `OpenAILLM` succeeds when `reasoning_effort="none"` is supplied, including a forced function call.

## Validation

- `pytest tests/core/model/chat/basic/test_openai.py -q` (68 passed)
- `ruff check src/xagent/core/model/chat/basic/openai.py tests/core/model/chat/basic/test_openai.py`
- `ruff format --check src/xagent/core/model/chat/basic/openai.py tests/core/model/chat/basic/test_openai.py`
- mutation check: the new regression test fails when the production change is reverted

### CODE REVIEW

File: src/xagent/core/model/chat/basic/openai.py [hunk 1]
```diff
@@ -186,6 +186,19 @@ def _degrade_rejected_params(
         completion_params.pop("response_format")
         degraded.append("response_format")
 
+    # Some OpenAI-compatible chat-completions endpoints enable reasoning by
+    # default, but reject that default when function tools are present. Their
+    # error explicitly asks clients to opt out, even when the original request
+    # did not contain ``reasoning_effort``.
+    if (
+        "tools" in completion_params
+        and rejected_param == "reasoning_effort"
```
Comment (rogercloud) [inline src/xagent/core/model/chat/basic/openai.py:195]: **Major, non-blocking.** This branch requires `rejected_param == "reasoning_effort"` with no fallback for when the endpoint's error body omits the `param` field (in which case `_openai_rejected_param` returns `None`). This is inconsistent with the two sibling branches in this same function: the `max_tokens` branch accepts `rejected_param is None or rejected_param == "max_tokens"`, and the `response_format` branch falls back to a text-only match when `rejected_param is None`. Suggest: `if (rejected_param is None or rejected_param == "reasoning_effort") and ...`.

Note: this exact gap was already flagged inline by the Gemini Code Assist bot on this PR (same suggestion) and appears unaddressed in the current diff -- raising it here for visibility, not as a second/duplicate finding.

Not blocking: it doesn't regress prior behavior (a provider omitting `param` simply falls through to the original re-raise, same as before this PR existed), but it does mean the fix may not fire for every gateway shape it's meant to handle.

File: src/xagent/core/model/chat/basic/openai.py [hunk 2]
```diff
@@ -186,6 +186,19 @@ def _degrade_rejected_params(
...
+        and "set reasoning_effort to 'none'" in lowered
```
Comment (rogercloud) [inline src/xagent/core/model/chat/basic/openai.py:196]: **Minor, non-blocking.** The text match `"set reasoning_effort to 'none'"` is single-quote-exact; a double-quoted or paraphrased variant of the same instruction (e.g. `set reasoning_effort to "none"`) would not match, even though `rejected_param == "reasoning_effort"` already reliably identifies the parameter. Given the strong structured `rejected_param` signal already gates this branch, this extra brittle text requirement adds fragility without much added precision, and is inconsistent with this file's own stated philosophy of trusting structured signals over rigid text matching (see the adjacent `response_format` branch's comment).

File: src/xagent/core/model/chat/basic/openai.py [hunk 3]
```diff
@@ -186,6 +186,19 @@ def _degrade_rejected_params(
...
+        and completion_params.get("reasoning_effort") != "none"
```
Comment (rogercloud) [inline src/xagent/core/model/chat/basic/openai.py:197]: **Minor, non-blocking.** If a caller (or `DeepSeekLLM`'s `DEEPSEEK_REASONING_EFFORT` config) had already set an explicit non-`"none"` `reasoning_effort`, this branch silently overrides it to `"none"` once the request has already 400'd. This is defensible last-resort recovery (the request already failed with that value, so retrying with the same value would only repeat the same 400), but the current comment only documents the "originally absent" scenario. Worth clarifying in the comment that both cases (absent, and explicit-but-rejected) are covered.

File: tests/core/model/chat/basic/test_openai.py [hunk 4]
```diff
@@ -1764,6 +1764,12 @@ def _bad_request(
...
 )
 
+_REASONING_TOOLS_REJECTION = (
+    "Function tools with reasoning_effort are not supported for this model in "
+    "/v1/chat/completions. To use function tools, use /v1/responses or set "
+    "reasoning_effort to 'none'."
+)
+
 
 class TestRejectedParameterDegrade:
```
Comment (rogercloud) [inline tests/core/model/chat/basic/test_openai.py:1774]: **Minor, non-blocking.** This class docstring still describes only the max_tokens/compaction scenario, but the class now also covers `response_format` and (with this PR) `reasoning_effort` degrades. Worth updating to describe the class's current scope.

File: tests/core/model/chat/basic/test_openai.py [hunk 5]
```diff
@@ -1799,6 +1805,46 @@ async def test_chat_renames_max_tokens(
...
+        llm = OpenAILLM(**openai_llm_config)
+        response = await llm.chat(
+            [{"role": "user", "content": "Call the tool."}], tools=tools
+        )
+
+        assert response["content"] == "Hello World"
+        first_kwargs = mock_client.chat.completions.create.call_args_list[0].kwargs
+        retry_kwargs = mock_client.chat.completions.create.call_args_list[1].kwargs
+        assert "reasoning_effort" not in first_kwargs
+        assert retry_kwargs["reasoning_effort"] == "none"
```
Comment (rogercloud) [inline tests/core/model/chat/basic/test_openai.py:1845]: **Minor, non-blocking.** This test only covers `chat()`; the sibling `max_tokens` degrade has parallel coverage across `chat()`, `vision_chat()`, and `stream_chat()` (`test_vision_chat_renames_max_tokens`, `test_stream_chat_renames_max_tokens_and_keeps_extra_body`), but no equivalent exists here for `reasoning_effort`. Also missing: a test for `rejected_param is None` with the same message text (would pin down the fallback gap noted at line 195 either way), a test for the tools-absent case (should NOT degrade / should re-raise), a test for an already-set non-`"none"` `reasoning_effort` being overridden, and a reasoning_effort-specific boundedness/loop-termination test (the existing one only covers the other two degrades). This test itself is well-constructed and was previously confirmed to fail when the production fix is reverted.

**Review by rogercloud — APPROVED** [review]
## Summary

This PR adds a third branch to `_degrade_rejected_params()` in `src/xagent/core/model/chat/basic/openai.py`, the shared 400-error retry helper used by `chat()`, `vision_chat()`, and `stream_chat()` across all OpenAI-compatible LLM subclasses (`OpenAILLM`, `AzureOpenAILLM`, `DashScopeLLM`, `OpenRouterLLM`, `DeepSeekLLM` -- all inherit it, none override it). Unlike the two existing branches, which only remove or rename an already-present parameter (`max_tokens` -> `max_completion_tokens`, or dropping `response_format`), this new branch *adds* `reasoning_effort="none"` to the retry when a 400 response rejects `reasoning_effort` on a tool-call request and instructs the caller to set it to `"none"` -- working around gateways (e.g. an AWS Bedrock OpenAI-compat shim) that reject function tools unless reasoning is explicitly disabled.

**Design verdict:** Acceptable-with-reservations. The fix is correctly placed in the single shared helper rather than duplicated per-subclass (verified: no subclass overrides or duplicates this retry loop). Its main reservations are that it's the first branch to *add* rather than remove/rename a parameter -- which breaks a documented invariant elsewhere in the file -- that it doesn't touch `extra_body`/`thinking` locals some subclasses also use for reasoning control (narrow edge cases only), and that its match conditions are stricter than its sibling branches in a way that may keep it from firing on some real gateway response shapes, somewhat undercutting the compatibility goal it exists for. None of the findings below regress prior behavior or are confirmed to break anything at runtime; all are minor/major-but-non-blocking.

Blocking: no -- recommended event: APPROVE

---

## Additional findings not anchored to the diff

GitHub only allows inline comments within the changed hunks. These findings point at pre-existing lines the diff doesn't touch, so they're listed here instead:

1. **`src/xagent/core/model/chat/basic/openai.py:568` (and equivalents at ~931, ~1181), vs. the new branch at line 199** -- Minor, non-blocking. The degrade only mutates `completion_params`; `extra_body` (built once before the retry loop via `_prepare_provider_reasoning_extra_body`) is never synced. On `DeepSeekLLM`/`OpenRouterLLM`, if a caller explicitly requested `thinking={"type": "enabled"}` together with `tools`, `extra_body` could still assert enabled reasoning while `completion_params["reasoning_effort"]` says `"none"` in the same retried request. Verified this is a narrow edge case (both subclasses only reach that `extra_body` state when the caller explicitly opts in), and the base `OpenAILLM` this PR targets never sets these `extra_body` keys at all, so it's unaffected there. Worth a follow-up to sync `extra_body` when this degrade fires.

2. **`src/xagent/core/model/chat/basic/openai.py:753`** -- Minor, non-blocking. Unlike the `response_format` degrade, which explicitly resets its local (`response_format = None`) after firing, the new branch doesn't reset the `thinking` local passed into `_process_response(... request_thinking=thinking)`. On `OpenRouterLLM`'s DeepSeek tool-protocol path this can produce a spurious "tool call without thinking disabled" WARNING log for a response where reasoning was in fact correctly disabled via the retry. Log-only impact -- no effect on the returned result or provider-state metadata.

3. **`src/xagent/core/model/chat/basic/openai.py:738` (and ~965, ~1217)** -- Minor, non-blocking. The shared log line `"API rejected %s, retrying without it. Error: %s"` is accurate for the two removal/rename degrades but backwards for the new branch, which *adds* `reasoning_effort` to the retry rather than omitting it. Purely a log-clarity issue -- not parsed or asserted anywhere in the codebase.

4. **`src/xagent/core/model/chat/basic/openai.py:138` (docstring) and the three "Bounded by construction" comments at ~746/~972/~1224; also `tests/core/model/chat/basic/test_openai.py:2234`** -- Minor, non-blocking. The function docstring ("Drop or rename request parameters...") and the near-identical "every degrade removes or renames a parameter, triggered only while it's still in the request" comments are no longer accurate for the new branch, which adds a parameter and fires while it's *absent*. The actual bound for this branch is a separate guard (`completion_params.get("reasoning_effort") != "none"`) that none of these comments describe. The boundedness test's docstring at test_openai.py:2234 likewise still describes only the original invariant. No runtime impact -- maintainability/accuracy only.

---

## Blocking status & recommended decision

Blocking: no. None of the findings meet the blocking bar -- no confirmed regression, no confirmed runtime break; all are either missing-fallback robustness gaps, narrow edge cases, log/comment accuracy issues, or test-coverage gaps.

- [new] Major, non-blocking -- missing `rejected_param is None` fallback (openai.py:195), duplicates an existing unaddressed Gemini Code Assist bot comment
- [new] Minor, non-blocking -- brittle single-quote-exact text match (openai.py:196)
- [new] Minor, non-blocking -- silent override of explicit non-`"none"` `reasoning_effort` undocumented (openai.py:197)
- [new] Minor, non-blocking -- stale class docstring (test_openai.py:1774)
- [new] Minor, non-blocking -- test coverage gap: only `chat()` covered (test_openai.py)
- [new] Minor, non-blocking -- `extra_body` not synced when degrade fires (see "Additional findings" above)
- [new] Minor, non-blocking -- `thinking` local not reset after degrade (see "Additional findings" above)
- [new] Minor, non-blocking -- shared log message inaccurate for the add-vs-remove case (see "Additional findings" above)
- [new] Minor, non-blocking -- stale docstrings/"Bounded by construction" comments (see "Additional findings" above)

Recommended event: APPROVE
